// @iwsdk/xr-input/dist/rig/xr-origin.js
import { Group, Matrix4 } from 'three';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class XROrigin extends Group {
    constructor() {
        super();
        this.raySpaces = {
            left: new Group(),
            right: new Group(),
        };
        this.gripSpaces = {
            left: new Group(),
            right: new Group(),
        };
        this.secondaryRaySpaces = {
            left: new Group(),
            right: new Group(),
        };
        this.secondaryGripSpaces = {
            left: new Group(),
            right: new Group(),
        };
        this.headsetMatrix = new Matrix4();
        this.head = new Group();
        this.head.name = 'xr-origin-head';
        this.add(this.head, this.raySpaces.left, this.raySpaces.right, this.gripSpaces.left, this.gripSpaces.right);
    }
    updateHead(frame, referenceSpace) {
        const pose = frame.getViewerPose(referenceSpace);
        if (pose) {
            this.headsetMatrix.fromArray(pose.transform.matrix);
            this.headsetMatrix.decompose(this.head.position, this.head.quaternion, this.head.scale);
        }
    }
}

export { XROrigin };
