// @iwsdk/xr-input/dist/gamepad/input-profiles.js
import { PROFILES_LIST, getProfile } from './generated-profiles.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const DEFAULT_PROFILES_PATH = 'https://cdn.jsdelivr.net/npm/@webxr-input-profiles/assets@1.0/dist/profiles';
const DEFAULT_PROFILE = 'generic-trigger';
const PROFILE_LIST_NAME = 'profilesList.json';
async function fetchJsonFile(path) {
    const response = await fetch(path);
    if (!response.ok) {
        throw new Error(response.statusText);
    }
    else {
        return response.json();
    }
}
async function fetchProfilesList(basePath = DEFAULT_PROFILES_PATH) {
    const profilesList = await fetchJsonFile(`${basePath}/${PROFILE_LIST_NAME}`);
    return profilesList;
}
function fetchProfileSync(inputSource, { defaultProfile = DEFAULT_PROFILE, } = {}) {
    const supportedProfilesList = PROFILES_LIST;
    let match;
    inputSource.profiles.some((profileId) => {
        const supportedProfile = supportedProfilesList[profileId];
        if (supportedProfile) {
            match = {
                profileId,
                profilePath: supportedProfile.path,
                deprecated: !!supportedProfile.deprecated,
            };
        }
        return !!match;
    });
    if (!match) {
        const supportedProfile = supportedProfilesList[defaultProfile];
        if (!supportedProfile) {
            throw new Error(`No matching profile name found and default profile "${defaultProfile}" missing.`);
        }
        match = {
            profileId: defaultProfile,
            profilePath: supportedProfile.path,
            deprecated: !!supportedProfile.deprecated,
        };
    }
    const profile = getProfile(match.profilePath);
    return profile;
}
async function fetchProfile(inputSource, { basePath = DEFAULT_PROFILES_PATH, defaultProfile = DEFAULT_PROFILE, }) {
    const supportedProfilesList = await fetchProfilesList(basePath);
    let match;
    inputSource.profiles.some((profileId) => {
        const supportedProfile = supportedProfilesList[profileId];
        if (supportedProfile) {
            match = {
                profileId,
                profilePath: `${basePath}/${supportedProfile.path}`,
                deprecated: !!supportedProfile.deprecated,
            };
        }
        return !!match;
    });
    if (!match) {
        const supportedProfile = supportedProfilesList[defaultProfile];
        if (!supportedProfile) {
            throw new Error(`No matching profile name found and default profile "${defaultProfile}" missing.`);
        }
        match = {
            profileId: defaultProfile,
            profilePath: `${basePath}/${supportedProfile.path}`,
            deprecated: !!supportedProfile.deprecated,
        };
    }
    const profile = await fetchJsonFile(match.profilePath);
    return profile;
}
function loadInputProfile(inputSource) {
    const profileId = inputSource.profiles[0];
    let profile = fetchProfileSync(inputSource);
    const layout = profile.layouts[inputSource.handedness];
    if (!layout) {
        throw new DOMException('No applicable layout found', 'NotSupportedError');
    }
    return {
        inputSource,
        layout,
        profileId,
        resolvedProfileId: profile.profileId,
        assetPath: `${DEFAULT_PROFILES_PATH}/${profile.profileId}/${layout.assetPath}`,
    };
}

export { DEFAULT_PROFILES_PATH, fetchJsonFile, fetchProfile, fetchProfileSync, fetchProfilesList, loadInputProfile };
