// @iwsdk/xr-input/dist/gamepad/stateful-gamepad.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var AxesState;
(function (AxesState) {
    AxesState[AxesState["Default"] = 0] = "Default";
    AxesState[AxesState["Up"] = 1] = "Up";
    AxesState[AxesState["Down"] = 2] = "Down";
    AxesState[AxesState["Left"] = 3] = "Left";
    AxesState[AxesState["Right"] = 4] = "Right";
})(AxesState || (AxesState = {}));
var InputComponent;
(function (InputComponent) {
    InputComponent["Trigger"] = "xr-standard-trigger";
    InputComponent["Squeeze"] = "xr-standard-squeeze";
    InputComponent["Touchpad"] = "xr-standard-touchpad";
    InputComponent["Thumbstick"] = "xr-standard-thumbstick";
    InputComponent["A_Button"] = "a-button";
    InputComponent["B_Button"] = "b-button";
    InputComponent["X_Button"] = "x-button";
    InputComponent["Y_Button"] = "y-button";
    InputComponent["Thumbrest"] = "thumbrest";
    InputComponent["Menu"] = "menu";
})(InputComponent || (InputComponent = {}));
class StatefulGamepad {
    constructor({ inputSource, layout }) {
        this.buttonMapping = new Map();
        this.axesMapping = new Map();
        this.axesThreshold = 0.8;
        this.axesStates = new Map();
        this.axes2DValues = new Map();
        this.axesValues = new Map();
        this.currentIndex = 1;
        this.previousIndex = 0;
        this.handedness = inputSource.handedness;
        this.gamepad = inputSource.gamepad;
        this.inputSource = inputSource;
        this.selectComponentId = layout.selectComponentId;
        const numButtons = this.gamepad.buttons.length;
        this.pressedArrs = [
            new Int8Array(numButtons).fill(0),
            new Int8Array(numButtons).fill(0),
        ];
        this.touchedArrs = [
            new Int8Array(numButtons).fill(0),
            new Int8Array(numButtons).fill(0),
        ];
        this.valueArrs = [
            new Float32Array(numButtons).fill(0),
            new Float32Array(numButtons).fill(0),
        ];
        Object.entries(layout.components).forEach(([id, config]) => {
            const buttonIdx = config.gamepadIndices['button'];
            if (buttonIdx !== undefined) {
                this.buttonMapping.set(id, buttonIdx);
            }
            if (config.type === 'thumbstick' || config.type === 'touchpad') {
                this.axesMapping.set(id, {
                    x: config.gamepadIndices.xAxis,
                    y: config.gamepadIndices.yAxis,
                });
                this.axesValues.set(id, { x: 0, y: 0 });
                this.axes2DValues.set(id, 0);
                this.axesStates.set(id, {
                    prev: AxesState.Default,
                    curr: AxesState.Default,
                });
            }
        });
    }
    update() {
        this.currentIndex = 1 - this.currentIndex;
        this.previousIndex = 1 - this.previousIndex;
        this.gamepad.buttons.forEach((button, idx) => {
            this.pressedArrs[this.currentIndex][idx] = button.pressed ? 1 : 0;
            this.touchedArrs[this.currentIndex][idx] = button.touched ? 1 : 0;
            this.valueArrs[this.currentIndex][idx] = button.value;
        });
        this.axesMapping.forEach(({ x: xIdx, y: yIdx }, id) => {
            const axesValue = this.axesValues.get(id);
            const axesState = this.axesStates.get(id);
            axesState.prev = axesState.curr;
            axesValue.x = this.gamepad.axes[xIdx];
            axesValue.y = this.gamepad.axes[yIdx];
            const { x, y } = axesValue;
            const value2D = Math.sqrt(x * x + y * y);
            this.axes2DValues.set(id, value2D);
            if (value2D < this.axesThreshold) {
                axesState.curr = AxesState.Default;
            }
            else {
                if (Math.abs(x) > Math.abs(y)) {
                    axesState.curr = x > 0 ? AxesState.Right : AxesState.Left;
                }
                else {
                    axesState.curr = y > 0 ? AxesState.Down : AxesState.Up;
                }
            }
        });
    }
    getButtonState(id, stateArr) {
        const idx = this.buttonMapping.get(id);
        return idx !== undefined ? stateArr[this.currentIndex][idx] : 0;
    }
    getButtonPressedByIdx(idx) {
        return !!this.pressedArrs[this.currentIndex][idx];
    }
    getButtonPressed(id) {
        return !!this.getButtonState(id, this.pressedArrs);
    }
    getButtonTouchedByIdx(idx) {
        return !!this.touchedArrs[this.currentIndex][idx];
    }
    getButtonTouched(id) {
        return !!this.getButtonState(id, this.touchedArrs);
    }
    getButtonValueByIdx(idx) {
        var _a;
        return (_a = this.valueArrs[this.currentIndex][idx]) !== null && _a !== void 0 ? _a : 0;
    }
    getButtonValue(id) {
        const idx = this.buttonMapping.get(id);
        return idx !== undefined ? this.valueArrs[this.currentIndex][idx] : 0;
    }
    getButtonDownByIdx(idx) {
        return ((this.pressedArrs[this.currentIndex][idx] &
            ~this.pressedArrs[this.previousIndex][idx]) !==
            0);
    }
    getButtonDown(id) {
        const idx = this.buttonMapping.get(id);
        return idx !== undefined
            ? (this.pressedArrs[this.currentIndex][idx] &
                ~this.pressedArrs[this.previousIndex][idx]) !==
                0
            : false;
    }
    getButtonUpByIdx(idx) {
        return ((~this.pressedArrs[this.currentIndex][idx] &
            this.pressedArrs[this.previousIndex][idx]) !==
            0);
    }
    getButtonUp(id) {
        const idx = this.buttonMapping.get(id);
        return idx !== undefined
            ? (~this.pressedArrs[this.currentIndex][idx] &
                this.pressedArrs[this.previousIndex][idx]) !==
                0
            : false;
    }
    getSelectStart() {
        return this.getButtonDown(this.selectComponentId);
    }
    getSelectEnd() {
        return this.getButtonUp(this.selectComponentId);
    }
    getSelecting() {
        return this.getButtonPressed(this.selectComponentId);
    }
    getAxesValues(id) {
        return this.axesValues.get(id);
    }
    getAxesState(id) {
        var _a;
        return (_a = this.axesStates.get(id)) === null || _a === void 0 ? void 0 : _a.curr;
    }
    get2DInputValue(id) {
        return this.axes2DValues.get(id);
    }
    getAxesEnteringState(id, state) {
        const axesState = this.axesStates.get(id);
        return axesState
            ? axesState.curr === state && axesState.prev !== state
            : false;
    }
    getAxesLeavingState(id, state) {
        const axesState = this.axesStates.get(id);
        return axesState
            ? axesState.curr !== state && axesState.prev === state
            : false;
    }
    getAxesEnteringUp(id) {
        return this.getAxesEnteringState(id, AxesState.Up);
    }
    getAxesEnteringDown(id) {
        return this.getAxesEnteringState(id, AxesState.Down);
    }
    getAxesEnteringLeft(id) {
        return this.getAxesEnteringState(id, AxesState.Left);
    }
    getAxesEnteringRight(id) {
        return this.getAxesEnteringState(id, AxesState.Right);
    }
    getAxesLeavingUp(id) {
        return this.getAxesLeavingState(id, AxesState.Up);
    }
    getAxesLeavingDown(id) {
        return this.getAxesLeavingState(id, AxesState.Down);
    }
    getAxesLeavingLeft(id) {
        return this.getAxesLeavingState(id, AxesState.Left);
    }
    getAxesLeavingRight(id) {
        return this.getAxesLeavingState(id, AxesState.Right);
    }
}

export { AxesState, InputComponent, StatefulGamepad };
