// @iwsdk/core/dist/visibility/visibility.js
import { createSystem } from '../ecs/system.js';
import '../ecs/world.js';
import { createComponent, Types } from 'elics';
import '../ecs/component-registry.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const Visibility = createComponent('Visibility', {
    isVisible: { type: Types.Boolean, default: true },
}, 'Component to control if an entity object is visible');
function attachToEntity(entity) {
    const object3D = entity.object3D;
    if (!object3D) {
        return;
    }
    Object.defineProperty(object3D, 'visible', {
        get: () => {
            return entity.getValue(Visibility, 'isVisible');
        },
        set: (value) => {
            entity.setValue(Visibility, 'isVisible', value);
        },
        enumerable: true,
        configurable: true,
    });
}
function detachFromEntity(entity) {
    const object3D = entity.object3D;
    if (!object3D) {
        return;
    }
    Object.defineProperty(object3D, 'visible', {
        value: object3D.visible,
        enumerable: true,
        configurable: true,
    });
}
class VisibilitySystem extends createSystem({
    visibility: { required: [Visibility] },
}) {
    init() {
        this.queries.visibility.subscribe('qualify', attachToEntity);
        this.queries.visibility.subscribe('disqualify', detachFromEntity);
    }
}

export { Visibility, VisibilitySystem };
