// @iwsdk/core/dist/index.js
import { VERSION as VERSION$1 } from 'elics';
export { ComponentRegistry, Entity, Types, createComponent, eq, ge, gt, isin, le, lt, ne, nin } from 'elics';
import { REVISION } from 'three';
export * from 'three';
import { VERSION } from './version.js';
import './runtime/three.js';
export { createSystem } from './ecs/system.js';
export { VisibilityState, World } from './ecs/world.js';
export { NullEntity } from './ecs/entity.js';
export { GLXFComponentRegistry } from './ecs/component-registry.js';
export { Transform, TransformSystem } from './transform/transform.js';
export { SyncedVector3 } from './transform/synced-vector3.js';
export { SyncedQuaternion } from './transform/synced-quaternion.js';
export { AssetManager, AssetType } from './asset/asset-manager.js';
export { CacheManager } from './asset/cache-manager.js';
export { SlideSystem } from './locomotion/slide.js';
export { TeleportSystem } from './locomotion/teleport.js';
export { TurnSystem, TurningMethod } from './locomotion/turn.js';
export { LocomotionEnvironment, LocomotionSystem } from './locomotion/locomotion.js';
export { Visibility, VisibilitySystem } from './visibility/visibility.js';
export { UIKitDocument } from './ui/document.js';
export { ColorSchemeType, PanelDocument, PanelUI, PanelUISystem } from './ui/ui.js';
export { FollowBehavior, FollowSystem, Follower } from './ui/follow.js';
export { ScreenSpace, ScreenSpaceUISystem } from './ui/screenspace.js';
export { ReferenceSpaceType, SessionMode, buildSessionInit, launchXR, normalizeReferenceSpec, resolveReferenceSpaceType } from './init/xr.js';
export { initializeWorld } from './init/world-initializer.js';
import './grab/index.js';
export { InputSystem } from './input/input-system.js';
export { Hovered, Interactable, Pressed } from './input/state-tags.js';
export { DomeTexture } from './environment/dome-texture.js';
export { DomeGradient } from './environment/dome-gradient.js';
export { IBLTexture } from './environment/ibl-texture.js';
export { IBLGradient } from './environment/ibl-gradient.js';
export { EnvironmentSystem } from './environment/environment-system.js';
export { LevelSystem } from './level/level-system.js';
export { LevelTag } from './level/level-tag.js';
export { GLXFImporter } from './level/level-glxf-importer.js';
export { EntityCreator } from './level/level-entity-creator.js';
export { LevelRoot } from './level/level-root.js';
export { SceneUnderstandingSystem } from './scene-understanding/scene-understanding-system.js';
export { XRAnchor } from './scene-understanding/anchor.js';
export { XRPlane } from './scene-understanding/plane.js';
export { XRMesh } from './scene-understanding/mesh.js';
export { AudioSource, DistanceModel, InstanceStealPolicy, PlaybackMode } from './audio/audio.js';
export { AudioSystem } from './audio/audio-system.js';
export { AudioUtils } from './audio/audio-utils.js';
export { AudioPool } from './audio/audio-pool.js';
export { PhysicsSystem } from './physics/physics-system.js';
export { DEFAULT_ANGULAR_DAMPING, DEFAULT_GRAVITY_FACTOR, DEFAULT_LINEAR_DAMPING, PhysicsBody, PhysicsState } from './physics/physicsBody.js';
export { PhysicsManipulation } from './physics/physicsManipulation.js';
export { DEFAULT_DENSITY, DEFAULT_FRICTION, DEFAULT_RESTITUTION, PhysicsShape, PhysicsShapeType } from './physics/physicsShape.js';
export { CameraFacing, CameraState } from './camera/types.js';
export { CameraUtils } from './camera/camera-utils.js';
export { CameraSource } from './camera/camera-source.js';
export { CameraSystem } from './camera/camera-system.js';
export * from '@iwsdk/xr-input';
export * from '@iwsdk/locomotor';
import * as uikit from '@pmndrs/uikit';
export { uikit as UIKit };
export { MovementMode } from './grab/handles.js';
export { GrabSystem } from './grab/grab-system.js';
export { OneHandGrabbable } from './grab/one-hand-grabbable.js';
export { TwoHandsGrabbable } from './grab/two-hands-grabbable.js';
export { DistanceGrabbable } from './grab/distance-grabbable.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
console.log(`                                                 
▄▄▄▄▄▄  ▄▄      ▄▄   ▄▄▄▄    ▄▄▄▄▄     ▄▄   ▄▄▄ 
▀▀██▀▀  ██      ██ ▄█▀▀▀▀█   ██▀▀▀██   ██  ██▀  
  ██    ▀█▄ ██ ▄█▀ ██▄       ██    ██  ██▄██    
  ██     ██ ██ ██   ▀████▄   ██    ██  █████    
  ██     ███▀▀███       ▀██  ██    ██  ██  ██▄  
▄▄██▄▄   ███  ███  █▄▄▄▄▄█▀  ██▄▄▄██   ██   ██▄ 
▀▀▀▀▀▀   ▀▀▀  ▀▀▀   ▀▀▀▀▀    ▀▀▀▀▀     ▀▀    ▀▀                                                
===============================================
Immersive Web SDK v${VERSION}
Three.js r${REVISION}
EliCS v${VERSION$1}`);

export { VERSION };
