// @iwsdk/core/dist/asset/cache-manager.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * Central cache for assets and in-flight loads used by {@link AssetManager}.
 *
 * @remarks
 * - Tracks a key→URL registry so callers may refer to assets by logical keys.
 * - De-duplicates concurrent requests via a promise cache.
 * - Stores resolved assets for fast reuse.
 *
 * @category Assets
 */
class CacheManager {
    /** Record a logical key → URL mapping. */
    static setKeyToUrl(key, url) {
        this.keyToUrl.set(key, url);
    }
    /** Resolve a key or pass-through a URL. */
    static resolveUrl(urlOrKey) {
        var _a;
        return (_a = this.keyToUrl.get(urlOrKey)) !== null && _a !== void 0 ? _a : urlOrKey;
    }
    /** True if an in-flight promise exists for URL. */
    static hasPromise(url) {
        return this.promiseCache.has(url);
    }
    /** Get the in-flight promise for URL (if any). */
    static getPromise(url) {
        return this.promiseCache.get(url);
    }
    /** Store an in-flight promise for URL. */
    static setPromise(url, promise) {
        this.promiseCache.set(url, promise);
    }
    /** Remove in-flight promise (on resolve/reject). */
    static deletePromise(url) {
        this.promiseCache.delete(url);
    }
    /** True if an asset has been cached for URL. */
    static hasAsset(url) {
        return this.cache.has(url);
    }
    /** Retrieve a cached asset by URL. */
    static getAsset(url) {
        return this.cache.get(url);
    }
    /** Store an asset by URL. */
    static setAsset(url, asset) {
        this.cache.set(url, asset);
    }
    /** Lookup asset by logical key or direct URL. */
    static getAssetByKey(key) {
        // First check if this is a URL (direct cache access)
        if (this.cache.has(key)) {
            return this.cache.get(key);
        }
        // Otherwise, look up the URL from the key registry
        const url = this.keyToUrl.get(key);
        if (url) {
            return this.cache.get(url);
        }
        return undefined;
    }
    /** Clear caches (useful during tests). */
    static clear() {
        this.cache.clear();
        this.promiseCache.clear();
        this.keyToUrl.clear();
    }
}
CacheManager.cache = new Map();
CacheManager.promiseCache = new Map();
CacheManager.keyToUrl = new Map();

export { CacheManager };
