// @iwsdk/xr-input/dist/visual/adapter/controller-visual-adapter.js
import { loadInputProfile } from '../../gamepad/input-profiles.js';
import { XRInputVisualAdapter } from './base-visual-adapter.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class XRControllerVisualAdapter extends XRInputVisualAdapter {
    constructor(playerSpace, handedness, visualsEnabled, visualClass, scene, camera, assetLoader) {
        super(playerSpace, handedness, visualsEnabled, visualClass, scene, camera, assetLoader);
    }
    connect(inputSource) {
        super.connect(inputSource);
        const inputConfig = loadInputProfile(inputSource);
        this.inputConfig = inputConfig;
        this.connectVisual();
    }
    disconnect() {
        super.disconnect();
        this.inputConfig = undefined;
        this.disconnectVisual();
    }
    update(_frame, delta) {
        if (this._inputSource) {
            if (this.visual && this.gripSpace) {
                this.visual.model.position.copy(this.gripSpace.position);
                this.visual.model.quaternion.copy(this.gripSpace.quaternion);
                this.visual.update(delta);
            }
        }
    }
    toggleVisual(enabled) {
        this.visualsEnabled = enabled;
        if (this.visual) {
            this.visual.toggle(enabled);
        }
    }
}

export { XRControllerVisualAdapter };
