// @iwsdk/locomotor/dist/core/movement-controller.js
import { Vector3 } from 'three';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * Handles movement input processing and velocity management
 * Reuses temporary vectors for performance
 */
class MovementController {
    constructor(config) {
        this.config = config;
        this.targetVelocity = new Vector3();
        this.velocityOnPlane = new Vector3();
        this.deltaVelocity = new Vector3();
        this.upAxis = new Vector3(0, 1, 0);
    }
    /**
     * Handle movement input and apply acceleration/deceleration
     * Modifies playerVelocity in-place for performance
     */
    handleMovementInput(movementVector, playerVelocity, isGrounded, delta) {
        if (movementVector.lengthSq() > 0) {
            this.targetVelocity.copy(movementVector);
            this.velocityOnPlane.copy(playerVelocity).projectOnPlane(this.upAxis);
            this.deltaVelocity.subVectors(this.targetVelocity, this.velocityOnPlane);
            // Apply acceleration with reduced control in air
            const accelFactor = isGrounded ? 1 : this.config.airDragFactor;
            const maxDeltaVel = this.config.acceleration * accelFactor * delta;
            this.deltaVelocity.clampLength(0, maxDeltaVel);
            playerVelocity.add(this.deltaVelocity);
        }
        else {
            this.velocityOnPlane.copy(playerVelocity).projectOnPlane(this.upAxis);
            const decelFactor = isGrounded ? 1 : 0.7;
            const maxDecel = this.config.deceleration * decelFactor * delta;
            this.deltaVelocity.copy(this.velocityOnPlane).clampLength(0, maxDecel);
            playerVelocity.sub(this.deltaVelocity);
        }
    }
    updateConfig(config) {
        Object.assign(this.config, config);
    }
}

export { MovementController };
