// @iwsdk/core/dist/ecs/system.js
import { signal } from '@preact/signals-core';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
/**
 * Create a strongly-typed System class with query bindings and reactive config.
 *
 * @param queries Elics query descriptors keyed by name.
 * @param schema Option map of config defaults and Types.
 * @returns A System constructor to `export class MySystem extends createSystem(...) { ... }`.
 *
 * @example
 * export class Rotator extends createSystem({ items: { required: [Transform] } }, {
 *   speed: { type: Types.Float32, default: 1 }
 * }) {
 *   update(dt:number){ this.queries.items.entities.forEach(e=> e.object3D.rotateY(dt*this.config.speed.value)) }
 * }
 *
 * @category ECS
 */
function createSystem(queries = {}, schema = {}) {
    var _a;
    return _a = class {
            constructor(world, queryManager, priority) {
                this.world = world;
                this.queryManager = queryManager;
                this.priority = priority;
                this.isPaused = false;
                this.config = {};
                this.cleanupFuncs = [];
                for (const key in schema) {
                    this.config[key] = signal(schema[key].default);
                }
                this.player = world.player;
                this.input = world.input;
                this.scene = world.scene;
                this.camera = world.camera;
                this.renderer = world.renderer;
                this.visibilityState = world.visibilityState;
            }
            get globals() {
                return this.world.globals;
            }
            get xrManager() {
                return this.world.renderer.xr;
            }
            get xrFrame() {
                return this.xrManager.getFrame();
            }
            createEntity() {
                return this.world.createEntity();
            }
            createTransformEntity(object, parent) {
                return this.world.createTransformEntity(object, parent);
            }
            init() { }
            update(_delta, _time) { }
            play() {
                this.isPaused = false;
            }
            stop() {
                this.isPaused = true;
            }
            destroy() {
                this.cleanupFuncs.forEach((func) => func());
            }
        },
        _a.schema = schema,
        _a.isSystem = true,
        _a.queries = queries,
        _a;
}

export { createSystem };
