// @iwsdk/xr-input/dist/visual/impl/animated-controller-hand.js
import { Matrix4, Vector3, BufferAttribute, Quaternion } from 'three';
import { lerp } from '../../node_modules/.pnpm/super-three@0.177.0/node_modules/super-three/src/math/MathUtils.js';
import { stencilMaterial, outlineMaterial } from './animated-hand.js';
import { BaseControllerVisual } from './base-impl.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const defaultPose = {
    wrist: [
        -0.19099761545658112, -0.5637121796607971, 0.8035844564437866, 0,
        0.9546268582344055, -0.29723578691482544, 0.01838778331875801, 0,
        0.22848893702030182, 0.7706354260444641, 0.5949063897132874, 0,
        0.058663930743932724, 0.03852901607751846, 0.07338987290859222, 1,
    ],
    'thumb-metacarpal': [
        0.7253018617630005, -0.6526158452033997, -0.21915610134601593, 0,
        0.5962181091308594, 0.7546267509460449, -0.273974746465683, 0,
        0.3441815674304962, 0.06804964691400528, 0.9364338517189026, 0,
        0.04146573320031166, 0.03385084494948387, 0.028252966701984406, 1,
    ],
    'thumb-phalanx-proximal': [
        0.6704757213592529, -0.7404858469963074, -0.04628881439566612, 0,
        0.6004704236984253, 0.5782264471054077, -0.5523486733436584, 0,
        0.4357720613479614, 0.3425414264202118, 0.8323269486427307, 0,
        0.030275389552116394, 0.03163835033774376, -0.0021932281088083982, 1,
    ],
    'thumb-phalanx-distal': [
        0.7001850605010986, -0.674904465675354, -0.23290452361106873, 0,
        0.4586094319820404, 0.6751725077629089, -0.5777708292007446, 0,
        0.5471910834312439, 0.2977343797683716, 0.7822635769844055, 0,
        0.015549290925264359, 0.020062798634171486, -0.030320117250084877, 1,
    ],
    'thumb-tip': [
        0.7001850605010986, -0.674904465675354, -0.23290452361106873, 0,
        0.4586094319820404, 0.6751725077629089, -0.5777708292007446, 0,
        0.5471910834312439, 0.2977343797683716, 0.7822635769844055, 0,
        0.003033810993656516, 0.012982223182916641, -0.050306063145399094, 1,
    ],
    'index-finger-metacarpal': [
        -0.19099761545658112, -0.5637121796607971, 0.8035844564437866, 0,
        0.9546268582344055, -0.29723578691482544, 0.01838778331875801, 0,
        0.22848893702030182, 0.7706354260444641, 0.5949063897132874, 0,
        0.04518073797225952, 0.024484416469931602, 0.035587772727012634, 1,
    ],
    'index-finger-phalanx-proximal': [
        -0.15332399308681488, -0.6449125409126282, 0.74871826171875, 0,
        0.9067131280899048, -0.393060564994812, -0.1528862863779068, 0,
        0.3928902745246887, 0.6554316878318787, 0.645016610622406, 0,
        0.03424347564578056, -0.019998569041490555, -0.002778385067358613, 1,
    ],
    'index-finger-phalanx-intermediate': [
        -0.19366276264190674, -0.6146624684333801, 0.7646464705467224, 0,
        0.6857759952545166, -0.642173707485199, -0.34252533316612244, 0,
        0.7015737295150757, 0.4580419659614563, 0.5458863377571106, 0,
        0.019342219457030296, -0.04485730454325676, -0.027242127805948257, 1,
    ],
    'index-finger-phalanx-distal': [
        -0.17719687521457672, -0.5709590315818787, 0.8016274571418762, 0,
        0.5550868511199951, -0.7305805683135986, -0.3976558446884155, 0,
        0.812699019908905, 0.37450963258743286, 0.44638875126838684, 0,
        0.002291433745995164, -0.055989429354667664, -0.0405091792345047, 1,
    ],
    'index-finger-tip': [
        -0.17719687521457672, -0.5709590315818787, 0.8016274571418762, 0,
        0.5550868511199951, -0.7305805683135986, -0.3976558446884155, 0,
        0.812699019908905, 0.37450963258743286, 0.44638875126838684, 0,
        -0.015366664156317711, -0.06528240442276001, -0.0506625697016716, 1,
    ],
    'middle-finger-metacarpal': [
        -0.19099761545658112, -0.5637121796607971, 0.8035844564437866, 0,
        0.9546268582344055, -0.29723578691482544, 0.01838778331875801, 0,
        0.22848893702030182, 0.7706354260444641, 0.5949063897132874, 0,
        0.044298235327005386, 0.016478098928928375, 0.049932993948459625, 1,
    ],
    'middle-finger-phalanx-proximal': [
        -0.16133272647857666, -0.43254467844963074, 0.8870605230331421, 0,
        0.8241896033287048, -0.5534587502479553, -0.1199769452214241, 0,
        0.5428470969200134, 0.7117499709129333, 0.4457901418209076, 0,
        0.03471161052584648, -0.033450815826654434, 0.015055419877171516, 1,
    ],
    'middle-finger-phalanx-intermediate': [
        -0.17571251094341278, -0.4131202697753906, 0.8935638070106506, 0,
        0.503139317035675, -0.8178667426109314, -0.27918484807014465, 0,
        0.8461533784866333, 0.40053096413612366, 0.3515668213367462, 0,
        0.01140881422907114, -0.06400411576032639, -0.004081025719642639, 1,
    ],
    'middle-finger-phalanx-distal': [
        -0.22404845058918, -0.32706698775291443, 0.9180572628974915, 0,
        0.3520073890686035, -0.905570387840271, -0.2367122620344162, 0,
        0.9087865352630615, 0.2701280415058136, 0.3180219233036041, 0,
        -0.011902349069714546, -0.07503857463598251, -0.013766520656645298, 1,
    ],
    'middle-finger-tip': [
        -0.22404845058918, -0.32706698775291443, 0.9180572628974915, 0,
        0.3520073890686035, -0.905570387840271, -0.2367122620344162, 0,
        0.9087865352630615, 0.2701280415058136, 0.3180219233036041, 0,
        -0.034258950501680374, -0.08291316777467728, -0.021691782400012016, 1,
    ],
    'ring-finger-metacarpal': [
        -0.19099761545658112, -0.5637121796607971, 0.8035844564437866, 0,
        0.9546268582344055, -0.29723578691482544, 0.01838778331875801, 0,
        0.22848893702030182, 0.7706354260444641, 0.5949063897132874, 0,
        0.04211178794503212, 0.005066493526101112, 0.0646386668086052, 1,
    ],
    'ring-finger-phalanx-proximal': [
        -0.1997598111629486, -0.2856025695800781, 0.9372975826263428, 0,
        0.8318477272987366, -0.5549431443214417, 0.00819024071097374, 0,
        0.5178080797195435, 0.7813250422477722, 0.3484334349632263, 0,
        0.028829503804445267, -0.03772619366645813, 0.03454011306166649, 1,
    ],
    'ring-finger-phalanx-intermediate': [
        -0.25824522972106934, -0.23653194308280945, 0.9366758465766907, 0,
        0.5364312529563904, -0.8414682149887085, -0.06459354609251022, 0,
        0.8034617900848389, 0.48578131198883057, 0.34418854117393494, 0,
        0.008636998012661934, -0.06819482147693634, 0.020952558144927025, 1,
    ],
    'ring-finger-phalanx-distal': [
        -0.3179231286048889, -0.23989734053611755, 0.9172642827033997, 0,
        0.3197873532772064, -0.9379004836082458, -0.13445626199245453, 0,
        0.8925586938858032, 0.25058284401893616, 0.3748965263366699, 0,
        -0.012713711708784103, -0.08110369741916656, 0.011806294322013855, 1,
    ],
    'ring-finger-tip': [
        -0.3179231286048889, -0.23989734053611755, 0.9172642827033997, 0,
        0.3197873532772064, -0.9379004836082458, -0.13445626199245453, 0,
        0.8925586938858032, 0.25058284401893616, 0.3748965263366699, 0,
        -0.0339939221739769, -0.0887695923447609, 0.0027068445924669504, 1,
    ],
    'pinky-finger-metacarpal': [
        -0.4810284674167633, -0.16032777726650238, 0.8619199991226196, 0,
        0.7997848987579346, -0.48294851183891296, 0.3565172255039215, 0,
        0.35910364985466003, 0.860845685005188, 0.360539972782135, 0,
        0.03749338537454605, 0.002106056548655033, 0.07142738997936249, 1,
    ],
    'pinky-finger-phalanx-proximal': [
        -0.2540414035320282, -0.07621645927429199, 0.9641854166984558, 0,
        0.8106723427772522, -0.5604919195175171, 0.1692887842655182, 0,
        0.5275158882141113, 0.8246449828147888, 0.20417501032352448, 0,
        0.0211003590375185, -0.03719118610024452, 0.05496629327535629, 1,
    ],
    'pinky-finger-phalanx-intermediate': [
        -0.283662885427475, 0.05269712582230568, 0.9574748277664185, 0,
        0.45776572823524475, -0.8699304461479187, 0.18349729478359222, 0,
        0.8426066040992737, 0.490350604057312, 0.22264425456523895, 0,
        0.004894857760518789, -0.06252463161945343, 0.04869394749403, 1,
    ],
    'pinky-finger-phalanx-distal': [
        -0.36656567454338074, 0.007802525535225868, 0.9303591847419739, 0,
        0.3588726222515106, -0.9213966131210327, 0.14912478625774384, 0,
        0.8583937287330627, 0.38854464888572693, 0.33495256304740906, 0,
        -0.012219619937241077, -0.0724843218922615, 0.04417172446846962, 1,
    ],
    'pinky-finger-tip': [
        -0.36656567454338074, 0.007802525535225868, 0.9303591847419739, 0,
        0.3588726222515106, -0.9213966131210327, 0.14912478625774384, 0,
        0.8583937287330627, 0.38854464888572693, 0.33495256304740906, 0,
        -0.030510887503623962, -0.08212456107139587, 0.03678080439567566, 1,
    ],
};
const pressedPose = {
    wrist: [
        -0.19326625764369965, -0.700115978717804, 0.6873756647109985, 0,
        0.9811022877693176, -0.13126900792121887, 0.14215001463890076, 0,
        -0.009290123358368874, 0.7018587589263916, 0.7122553586959839, 0,
        0.04728994518518448, 0.03910332918167114, 0.07807964831590652, 1,
    ],
    'thumb-metacarpal': [
        0.836667537689209, -0.3977629840373993, -0.37652623653411865, 0,
        0.31428012251853943, 0.9116700291633606, -0.26473715901374817, 0,
        0.4485703110694885, 0.1031622588634491, 0.8877738118171692, 0,
        0.03273596242070198, 0.0349138081073761, 0.0306982584297657, 1,
    ],
    'thumb-phalanx-proximal': [
        0.8362306952476501, -0.5063737034797668, -0.21048520505428314, 0,
        0.29973605275154114, 0.743484377861023, -0.5978203415870667, 0,
        0.45921289920806885, 0.4368256628513336, 0.7735030055046082, 0,
        0.01815161481499672, 0.03155970200896263, 0.0018341526156291366, 1,
    ],
    'thumb-phalanx-distal': [
        0.8111386299133301, -0.4369579553604126, -0.3887442946434021, 0,
        0.08336401730775833, 0.744280219078064, -0.6626443862915039, 0,
        0.5788822174072266, 0.5050888657569885, 0.6401411294937134, 0,
        0.002633400959894061, 0.016797985881567, -0.02430492267012596, 1,
    ],
    'thumb-tip': [
        0.8111386299133301, -0.4369579553604126, -0.3887442946434021, 0,
        0.08336401730775833, 0.744280219078064, -0.6626443862915039, 0,
        0.5788822174072266, 0.5050888657569885, 0.6401411294937134, 0,
        -0.010972380638122559, 0.00484889093786478, -0.04098760336637497, 1,
    ],
    'index-finger-metacarpal': [
        -0.19326625764369965, -0.700115978717804, 0.6873756647109985, 0,
        0.9811022877693176, -0.13126900792121887, 0.14215001463890076, 0,
        -0.009290123358368874, 0.7018587589263916, 0.7122553586959839, 0,
        0.039665982127189636, 0.027023138478398323, 0.03726854547858238, 1,
    ],
    'index-finger-phalanx-proximal': [
        -0.11402399092912674, -0.6228838562965393, 0.773959755897522, 0,
        0.8602886199951172, -0.45154619216918945, -0.23666246235370636, 0,
        0.49689218401908875, 0.6388437151908875, 0.5873472690582275, 0,
        0.045555129647254944, -0.010823833756148815, -0.00752238230779767, 1,
    ],
    'index-finger-phalanx-intermediate': [
        -0.15015903115272522, -0.5908305048942566, 0.7926985621452332, 0,
        0.07557929307222366, -0.806303083896637, -0.5866537690162659, 0,
        0.9857685565948486, -0.028179632499814034, 0.16572844982147217, 0,
        0.02670934423804283, -0.03505346551537514, -0.029798904433846474, 1,
    ],
    'index-finger-phalanx-distal': [
        -0.10268659889698029, -0.5726102590560913, 0.8133710026741028, 0,
        -0.3890283405780792, -0.7294459939002991, -0.5626415014266968, 0,
        0.9154847860336304, -0.37420010566711426, -0.14785714447498322, 0,
        0.0027515734545886517, -0.034368596971035004, -0.03382670134305954, 1,
    ],
    'index-finger-tip': [
        -0.10268659889698029, -0.5726102590560913, 0.8133710026741028, 0,
        -0.3890283405780792, -0.7294459939002991, -0.5626415014266968, 0,
        0.9154847860336304, -0.37420010566711426, -0.14785714447498322, 0,
        -0.018150901421904564, -0.0269172266125679, -0.030856406316161156, 1,
    ],
    'middle-finger-metacarpal': [
        -0.19326625764369965, -0.700115978717804, 0.6873756647109985, 0,
        0.9811022877693176, -0.13126900792121887, 0.14215001463890076, 0,
        -0.009290123358368874, 0.7018587589263916, 0.7122553586959839, 0,
        0.039410170167684555, 0.01680479384958744, 0.05037049949169159, 1,
    ],
    'middle-finger-phalanx-proximal': [
        -0.19697882235050201, -0.5661503672599792, 0.8004202246665955, 0,
        0.44485533237457275, -0.7791463732719421, -0.44162672758102417, 0,
        0.8736720085144043, 0.269080251455307, 0.4053305983543396, 0,
        0.046016983687877655, -0.026484908536076546, 0.00840697530657053, 1,
    ],
    'middle-finger-phalanx-intermediate': [
        -0.19850671291351318, -0.543941080570221, 0.815305233001709, 0,
        -0.8491258025169373, -0.31999102234840393, -0.4202273488044739, 0,
        0.4894692301750183, -0.7757146954536438, -0.39835408329963684, 0,
        0.00851286482065916, -0.03803572431206703, -0.008992661722004414, 1,
    ],
    'middle-finger-phalanx-distal': [
        -0.02412884309887886, -0.5203070044517517, 0.8536380529403687, 0,
        -0.8494946360588074, 0.46083885431289673, 0.2568773031234741, 0,
        -0.527044951915741, -0.7189629077911377, -0.4531177282333374, 0,
        -0.004971807822585106, -0.016665101051330566, 0.0019818395376205444, 1,
    ],
    'middle-finger-tip': [
        -0.02412884309887886, -0.5203070044517517, 0.8536380529403687, 0,
        -0.8494946360588074, 0.46083885431289673, 0.2568773031234741, 0,
        -0.527044951915741, -0.7189629077911377, -0.4531177282333374, 0,
        0.007212246768176556, 0.0016472793649882078, 0.013849507085978985, 1,
    ],
    'ring-finger-metacarpal': [
        -0.19326625764369965, -0.700115978717804, 0.6873756647109985, 0,
        0.9811022877693176, -0.13126900792121887, 0.14215001463890076, 0,
        -0.009290123358368874, 0.7018587589263916, 0.7122553586959839, 0,
        0.03881341218948364, 0.0049891178496181965, 0.06276080012321472, 1,
    ],
    'ring-finger-phalanx-proximal': [
        -0.2744467258453369, -0.5059019923210144, 0.8177663087844849, 0,
        0.2607558071613312, -0.8577102422714233, -0.44310176372528076, 0,
        0.92557293176651, 0.09162963181734085, 0.36731287837028503, 0,
        0.038332559168338776, -0.03451777994632721, 0.025984058156609535, 1,
    ],
    'ring-finger-phalanx-intermediate': [
        -0.27134573459625244, -0.4247802495956421, 0.863673985004425, 0,
        -0.9208118319511414, -0.14660659432411194, -0.3614026606082916, 0,
        0.28013697266578674, -0.8933463096618652, -0.3513617217540741, 0,
        0.0022388119250535965, -0.03809098154306412, 0.011660284362733364, 1,
    ],
    'ring-finger-phalanx-distal': [
        -0.2017165720462799, -0.39981958270072937, 0.8941220641136169, 0,
        -0.6362114548683167, 0.7475619316101074, 0.1907520294189453, 0,
        -0.7446783781051636, -0.5303730368614197, -0.40516555309295654, 0,
        -0.005205388180911541, -0.014351737685501575, 0.02099715732038021, 1,
    ],
    'ring-finger-tip': [
        -0.2017165720462799, -0.39981958270072937, 0.8941220641136169, 0,
        -0.6362114548683167, 0.7475619316101074, 0.1907520294189453, 0,
        -0.7446783781051636, -0.5303730368614197, -0.40516555309295654, 0,
        0.011834603734314442, -3507334040477872e-19, 0.031390633434057236, 1,
    ],
    'pinky-finger-metacarpal': [
        -0.5599007606506348, -0.3645467460155487, 0.7440539598464966, 0,
        0.8176640272140503, -0.3882564902305603, 0.42506730556488037, 0,
        0.1339270919561386, 0.8463817834854126, 0.5154619812965393, 0,
        0.033919818699359894, 0.000323370419209823, 0.06828021258115768, 1,
    ],
    'pinky-finger-phalanx-proximal': [
        -0.4415345788002014, -0.4293745458126068, 0.7878351807594299, 0,
        0.2647094428539276, -0.9013156294822693, -0.3428681790828705, 0,
        0.8573074340820312, 0.05715937167406082, 0.5116217732429504, 0,
        0.027806390076875687, -0.03831324353814125, 0.04474702104926109, 1,
    ],
    'pinky-finger-phalanx-intermediate': [
        -0.3579302430152893, -0.30541327595710754, 0.8823878169059753, 0,
        -0.8971001505851746, -0.1496744304895401, -0.4157034456729889, 0,
        0.25903216004371643, -0.9403831958770752, -0.22041329741477966, 0,
        0.0014695478603243828, -0.04006920009851456, 0.029029792174696922, 1,
    ],
    'pinky-finger-phalanx-distal': [
        -0.33868440985679626, -0.2318958044052124, 0.9118753671646118, 0,
        -0.5754091143608093, 0.8178455233573914, -0.005732276476919651, 0,
        -0.7444442510604858, -0.5266430377960205, -0.41042670607566833, 0,
        -0.0037917618174105883, -0.020968729630112648, 0.033506691455841064, 1,
    ],
    'pinky-finger-tip': [
        -0.33868440985679626, -0.2318958044052124, 0.9118753671646118, 0,
        -0.5754091143608093, 0.8178455233573914, -0.005732276476919651, 0,
        -0.7444442510604858, -0.5266430377960205, -0.41042670607566833, 0,
        0.011911955662071705, -0.00837172381579876, 0.042272504419088364, 1,
    ],
};
class AnimatedControllerHand extends BaseControllerVisual {
    constructor(model, scene, camera, layout) {
        var _a;
        super(scene, camera, model, layout);
        this.bones = [];
        this.mat4 = new Matrix4();
        this.vec3 = new Vector3();
        const skinnedMesh = model.getObjectByProperty('type', 'SkinnedMesh');
        skinnedMesh.frustumCulled = false;
        skinnedMesh.parent.scale.multiplyScalar(1.05);
        const wrist = model.getObjectByName('wrist');
        const geometry = skinnedMesh.geometry;
        // Compute the wrist's position in the skinnedMesh’s local space.
        const wristWorldPos = wrist.getWorldPosition(new Vector3());
        const wristLocal = skinnedMesh.worldToLocal(wristWorldPos.clone());
        // Access the vertex positions.
        const posAttr = geometry.attributes.position;
        const vertexCount = posAttr.count;
        const distances = new Float32Array(vertexCount);
        for (let i = 0; i < vertexCount; i++) {
            const x = posAttr.getX(i);
            const y = posAttr.getY(i);
            const z = posAttr.getZ(i);
            // Compute distance from the vertex (in bind space) to the wristLocal position.
            distances[i] = Math.sqrt((x - wristLocal.x) * (x - wristLocal.x) +
                (y - wristLocal.y) * (y - wristLocal.y) +
                (z - wristLocal.z) * (z - wristLocal.z));
        }
        // Add the custom attribute to the geometry.
        geometry.setAttribute('wristDist', new BufferAttribute(distances, 1));
        skinnedMesh.material = stencilMaterial;
        const skinnedOutline = skinnedMesh.clone();
        skinnedOutline.material = outlineMaterial;
        (_a = skinnedMesh.parent) === null || _a === void 0 ? void 0 : _a.add(skinnedOutline);
        skinnedOutline.renderOrder = 999;
        Object.keys(AnimatedControllerHand.defaultPose).forEach((jointName) => {
            const bone = this.model.getObjectByName(jointName);
            bone.userData = {
                defaultTransform: {
                    position: new Vector3(),
                    quaternion: new Quaternion(),
                },
                pressedTransform: {
                    position: new Vector3(),
                    quaternion: new Quaternion(),
                },
                buttonIndex: jointName.startsWith('index')
                    ? 0
                    : jointName.startsWith('thumb')
                        ? 3
                        : 1,
                jointName,
                alpha: 0,
            };
            this.bones.push(bone);
        });
    }
    connect(inputSource, enabled) {
        super.connect(inputSource, enabled);
        this.bones.forEach((bone) => {
            const { jointName } = bone.userData;
            this.mat4.fromArray(AnimatedControllerHand.defaultPose[jointName]);
            this.mat4.decompose(bone.userData.defaultTransform.position, bone.userData.defaultTransform.quaternion, this.vec3);
            this.mat4.fromArray(AnimatedControllerHand.pressedPose[jointName]);
            this.mat4.decompose(bone.userData.pressedTransform.position, bone.userData.pressedTransform.quaternion, this.vec3);
            if (inputSource.handedness === 'left') {
                bone.userData.defaultTransform.position.x *= -1;
                bone.userData.pressedTransform.position.x *= -1;
                bone.userData.defaultTransform.quaternion.x *= -1;
                bone.userData.defaultTransform.quaternion.w *= -1;
                bone.userData.pressedTransform.quaternion.x *= -1;
                bone.userData.pressedTransform.quaternion.w *= -1;
            }
        });
    }
    update(delta) {
        if (this.gamepad && this.enabled) {
            this.bones.forEach((bone) => {
                if (bone) {
                    const { buttonIndex, defaultTransform, pressedTransform, alpha } = bone.userData;
                    const button = this.gamepad.buttons[buttonIndex];
                    const newAlpha = lerp(alpha, button.value, delta * 20);
                    bone.position.lerpVectors(defaultTransform.position, pressedTransform.position, newAlpha);
                    bone.quaternion.slerpQuaternions(defaultTransform.quaternion, pressedTransform.quaternion, newAlpha);
                    bone.userData.alpha = newAlpha;
                }
            });
        }
    }
}
AnimatedControllerHand.assetKeyPrefix = 'controller-hand-';
AnimatedControllerHand.assetProfileId = 'generic-hand';
AnimatedControllerHand.defaultPose = defaultPose;
AnimatedControllerHand.pressedPose = pressedPose;

export { AnimatedControllerHand };
