// @iwsdk/xr-input/dist/visual/impl/animated-controller.js
import { MeshMatcapMaterial } from 'three';
import { FlexBatchedMesh } from '../utils/flex-batched-mesh.js';
import { BaseControllerVisual } from './base-impl.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class AnimatedController extends BaseControllerVisual {
    constructor() {
        super(...arguments);
        this.animatedComponents = [];
    }
    init() {
        if (AnimatedController.useSimpleMaterial) {
            const simpleMaterials = new Map();
            this.model.traverse((node) => {
                const mesh = node;
                if (mesh.isMesh) {
                    const originalMat = mesh.material;
                    let simpleMat = simpleMaterials.get(originalMat);
                    if (!simpleMat) {
                        simpleMat = new MeshMatcapMaterial({
                            color: originalMat.color,
                            map: originalMat.map,
                            normalMap: originalMat.normalMap,
                        });
                        simpleMaterials.set(originalMat, simpleMat);
                    }
                    mesh.material = simpleMat;
                }
            });
        }
        Object.values(this.layout.components).forEach((config) => {
            Object.values(config.visualResponses).forEach((animConfig) => {
                const type = animConfig.componentProperty;
                if (type === 'button' || type === 'xAxis' || type === 'yAxis') {
                    const gamepadIndex = config.gamepadIndices[type];
                    const isButton = type === 'button';
                    const node = this.model.getObjectByName(animConfig.valueNodeName);
                    const minNode = animConfig.minNodeName
                        ? this.model.getObjectByName(animConfig.minNodeName)
                        : undefined;
                    const maxNode = animConfig.maxNodeName
                        ? this.model.getObjectByName(animConfig.maxNodeName)
                        : undefined;
                    if (gamepadIndex === undefined || !node || !minNode || !maxNode) {
                        return;
                    }
                    this.animatedComponents.push({
                        isButton,
                        gamepadIndex,
                        node,
                        transformRange: {
                            min: {
                                position: minNode.position.clone(),
                                quaternion: minNode.quaternion.clone(),
                            },
                            max: {
                                position: maxNode.position.clone(),
                                quaternion: maxNode.quaternion.clone(),
                            },
                        },
                    });
                    minNode.removeFromParent();
                    maxNode.removeFromParent();
                }
            });
        });
        this.model = new FlexBatchedMesh(this.model);
    }
    update() {
        if (this.enabled && this.gamepad) {
            this.animatedComponents.forEach((animComponent) => {
                const { isButton, node, gamepadIndex, transformRange } = animComponent;
                const alpha = isButton
                    ? this.gamepad.buttons[gamepadIndex].value
                    : (this.gamepad.axes[gamepadIndex] + 1) / 2;
                node.position.lerpVectors(transformRange.min.position, transformRange.max.position, alpha);
                node.quaternion.slerpQuaternions(transformRange.min.quaternion, transformRange.max.quaternion, alpha);
            });
        }
    }
}
AnimatedController.assetKeyPrefix = 'controller-';
AnimatedController.useSimpleMaterial = false;

export { AnimatedController };
