// @iwsdk/xr-input/dist/visual/impl/base-impl.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
class BaseControllerVisual {
    constructor(scene, camera, model, layout) {
        this.scene = scene;
        this.camera = camera;
        this.model = model;
        this.layout = layout;
        this.enabled = true;
    }
    init() { }
    connect(inputSource, enabled) {
        this.toggle(enabled);
        this.gamepad = inputSource.gamepad;
    }
    disconnect() {
        this.gamepad = undefined;
    }
    toggle(enabled) {
        if (this.enabled !== enabled) {
            this.model.visible = enabled;
            this.enabled = enabled;
        }
    }
    update(_delta) { }
}
class BaseHandVisual {
    constructor(scene, camera, model, layout) {
        this.scene = scene;
        this.camera = camera;
        this.model = model;
        this.layout = layout;
        this.joints = [];
        this.enabled = true;
    }
    init() { }
    connect(inputSource, enabled) {
        this.toggle(enabled);
        const hand = inputSource.hand;
        hand.forEach((jointSpace) => {
            const jointName = jointSpace.jointName;
            this.joints.push(this.model.getObjectByName(jointName));
        });
    }
    disconnect() {
        this.joints.length = 0;
    }
    toggle(enabled) {
        if (this.enabled !== enabled) {
            this.model.visible = enabled;
            this.enabled = enabled;
        }
    }
    update() {
        var _a;
        const jointTransforms = (_a = this.xrInput) === null || _a === void 0 ? void 0 : _a.jointTransforms;
        if (this.enabled && jointTransforms) {
            this.joints.forEach((bone, index) => {
                if (bone) {
                    bone.matrix.fromArray(jointTransforms, index * 16);
                    bone.matrix.decompose(bone.position, bone.quaternion, bone.scale);
                }
            });
        }
    }
}

export { BaseControllerVisual, BaseHandVisual };
