// @pmndrs/uikitml/dist/interpreter/html.js
import { Container, Input } from '@pmndrs/uikit';
export const htmlKit = {
    h1: class H1 extends Container {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: {
                    fontSize: 24,
                    fontWeight: 'bold',
                    ...config?.defaultOverrides,
                },
            });
        }
    },
    h2: class H2 extends Container {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: {
                    fontSize: 24,
                    fontWeight: 'bold',
                    ...config?.defaultOverrides,
                },
            });
        }
    },
    h3: class H3 extends Container {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: {
                    fontSize: 18.72,
                    fontWeight: 'bold',
                    ...config?.defaultOverrides,
                },
            });
        }
    },
    h4: class H4 extends Container {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: { fontSize: 16, fontWeight: 'bold', ...config?.defaultOverrides },
            });
        }
    },
    h5: class H5 extends Container {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: { fontSize: 13.28, fontWeight: 'bold', ...config?.defaultOverrides },
            });
        }
    },
    h6: class H6 extends Container {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: {
                    //tailwind disables this marginY: 37.28,
                    fontSize: 10.67,
                    fontWeight: 'bold',
                    ...config?.defaultOverrides,
                },
            });
        }
    },
    ol: class OL extends Container {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: { flexDirection: 'column', ...config?.defaultOverrides },
            });
        }
    },
    ul: class UL extends Container {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: { flexDirection: 'column', ...config?.defaultOverrides },
            });
        }
    },
    a: class A extends Container {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: {
                    cursor: 'pointer',
                    onClick: (e) => {
                        e.stopPropagation?.();
                        const href = this.properties.peek().href;
                        if (href != null) {
                            window.open(href);
                        }
                    },
                    ...config?.defaultOverrides,
                },
            });
        }
    },
    button: class Button extends Container {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: {
                    verticalAlign: 'middle',
                    textAlign: 'center',
                    cursor: 'pointer',
                    ...config?.defaultOverrides,
                },
            });
        }
    },
    textarea: class Textarea extends Input {
        constructor(inputProperties, initialClasses, config) {
            super(inputProperties, initialClasses, {
                ...config,
                defaultOverrides: { multiline: true, ...config?.defaultOverrides },
            });
        }
    },
};
//TBD select option
