// @iwsdk/core/dist/grab/index.js
import { RayIntersector, SphereIntersector } from '@pmndrs/pointer-events';
import { patchedIntersectRayPointerCapture, patchedIntersectSpherePointerCapture, patchedExecuteIntersection } from './pointer-events-patch.js';
import '../input/input-system.js';
export { Hovered, Interactable, Pressed } from '../input/state-tags.js';
export { GrabSystem } from './grab-system.js';
export { OneHandGrabbable } from './one-hand-grabbable.js';
export { TwoHandsGrabbable } from './two-hands-grabbable.js';
export { DistanceGrabbable } from './distance-grabbable.js';
export { MovementMode } from './handles.js';

/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// Apply monkeypatch once
RayIntersector.prototype.intersectPointerCapture =
    patchedIntersectRayPointerCapture;
SphereIntersector.prototype.intersectPointerCapture =
    patchedIntersectSpherePointerCapture;
SphereIntersector.prototype.executeIntersection = patchedExecuteIntersection;
