// @iwsdk/locomotor/dist/types/message-types.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var MessageType;
(function (MessageType) {
    MessageType[MessageType["Init"] = 0] = "Init";
    MessageType[MessageType["Config"] = 1] = "Config";
    MessageType[MessageType["Slide"] = 2] = "Slide";
    MessageType[MessageType["Teleport"] = 3] = "Teleport";
    MessageType[MessageType["ParabolicRaycast"] = 4] = "ParabolicRaycast";
    MessageType[MessageType["PositionUpdate"] = 5] = "PositionUpdate";
    MessageType[MessageType["RaycastUpdate"] = 6] = "RaycastUpdate";
    MessageType[MessageType["AddEnvironment"] = 7] = "AddEnvironment";
    MessageType[MessageType["RemoveEnvironment"] = 8] = "RemoveEnvironment";
    MessageType[MessageType["UpdateKinematicEnvironment"] = 9] = "UpdateKinematicEnvironment";
    MessageType[MessageType["Jump"] = 10] = "Jump";
})(MessageType || (MessageType = {}));

export { MessageType };
